From a924fb84d5ea51ca6a7f801c63b68cf9c1f080c4 Mon Sep 17 00:00:00 2001
From: Hannes Matuschek <hmatuschek@gmail.com>
Date: Wed, 22 Jan 2025 22:21:34 +0100
Subject: [PATCH] Fixed encoding/decoding of CTCSS & DCS subtones for OpenGD77
 (once again). Fixes #549. (#555)

* Updated regression test.
* Fixed missing leading digit.
---
 lib/opengd77base_codeplug.cc | 8 ++++++--
 test/opengd77_test.cc        | 4 ++--
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/lib/opengd77base_codeplug.cc b/lib/opengd77base_codeplug.cc
index 3f6411b..67d7fdd 100644
--- a/lib/opengd77base_codeplug.cc
+++ b/lib/opengd77base_codeplug.cc
@@ -40,7 +40,8 @@ OpenGD77BaseCodeplug::encodeSelectiveCall(const SelectiveCall &call) {
     toneCode = call.mHz()/100;
   }
 
-  uint16_t bcd = (((toneCode/100) % 10 ) << 8) |
+  uint16_t bcd = (((toneCode/1000) % 10 ) << 12) |
+      (((toneCode/100) % 10 ) << 8) |
       (((toneCode/10) % 10 ) << 4) |
       (((toneCode/1) % 10 )  << 0);
   return (dcs<<15) | (inverted << 14) | (bcd & 0x3fff);
@@ -55,7 +56,10 @@ OpenGD77BaseCodeplug::decodeSelectiveCall(uint16_t code) {
       inverted = ((code >> 14) & 1);
 
   uint16_t bcd = (code & 0x3fff);
-  code = 100 * ((bcd >> 8) & 0xf) + 10 * ((bcd >> 4) & 0xf) + 1 * ((bcd >> 0) & 0xf);
+  code = 1000 * ((bcd >> 12) & 0xf)
+      + 100 * ((bcd >> 8) & 0xf)
+      + 10 * ((bcd >> 4) & 0xf)
+      + 1 * ((bcd >> 0) & 0xf);
 
   if (! dcs)
     return SelectiveCall(double(code)/10);
diff --git a/test/opengd77_test.cc b/test/opengd77_test.cc
index 9af6d98..44fa231 100644
--- a/test/opengd77_test.cc
+++ b/test/opengd77_test.cc
@@ -149,7 +149,7 @@ OpenGD77Test::testChannelSubTones() {
   }
 
   config.channelList()->channel(0)->as<FMChannel>()->setTXTone(SelectiveCall(67.0));
-  config.channelList()->channel(0)->as<FMChannel>()->setRXTone(SelectiveCall(71.9));
+  config.channelList()->channel(0)->as<FMChannel>()->setRXTone(SelectiveCall(123.0));
 
   if (! encodeDecode(config, decoded, err))
     QFAIL(err.format().toLocal8Bit().constData());
@@ -163,7 +163,7 @@ OpenGD77Test::testChannelSubTones() {
 
   QVERIFY(rxTone.isValid());
   QVERIFY(rxTone.isCTCSS());
-  QCOMPARE(rxTone.Hz(), 71.9);
+  QCOMPARE(rxTone.Hz(), 123.0);
 }
 
 QTEST_GUILESS_MAIN(OpenGD77Test)
-- 
2.49.0

